### ISSUE TITLE
Build isn't reproducible due to gendynapi.pl

### ISSUE THREAD

**Issue by cpaelzer** [libsdl-org/SDL#11565]
Hi,
while debugging an issue that eventually turned out an [issue with LTO](https://bugs.launchpad.net/ubuntu/+source/libsdl2/+bug/2089779) I came by ensuring the builds are comparable for my debugging. That was made harder by the build being not reproducible as also [noted in the Debian project](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/libsdl2.html).

Delta in the objects was like:

```
-good/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9: file format elf64-x86-64
+bad/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9: file format elf64-x86-64

 Disassembly of section .init:
@@ -2333,7 +2333,7 @@
    32050: 48 8b 04 25 00 00 00 mov 0x0,%rax
    32057: 00
    32058: 0f 0b ud2
- 3205a: 48 8d 3d 23 90 1f 00 lea 0x1f9023(%rip),%rdi # 22b084 <SDL_DYNAPI_entry@@Base+0x1c72c4>
+ 3205a: 48 8d 3d 13 90 1f 00 lea 0x1f9013(%rip),%rdi # 22b074 <SDL_DYNAPI_entry@@Base+0x1c72b4>
    32061: e8 8a f5 ff ff call 315f0 <perror@plt>
    32066: e9 10 de 03 00 jmp 6fe7b <SDL_DYNAPI_entry@@Base+0xc0bb>
    3206b: 49 8b 74 24 08 mov 0x8(%r12),%rsi
@@ -2351,56 +2351,56 @@
    32098: c6 04 25 00 00 00 00 movb $0x0,0x0
    3209f: 00
    320a0: 0f 0b ud2
- 320a2: 48 8d 3d 15 b8 1f 00 lea 0x1fb815(%rip),%rdi # 22d8be <SDL_DYNAPI_entry@@Base+0x1c9afe>
+ 320a2: 48 8d 3d 05 b8 1f 00 lea 0x1fb805(%rip),%rdi # 22d8ae <SDL_DYNAPI_entry@@Base+0x1c9aee>
    320a9: e8 42 f5 ff ff call 315f0 <perror@plt>
- 320ae: 48 8d 3d 23 d6 1d 00 lea 0x1dd623(%rip),%rdi # 20f6d8 <SDL_DYNAPI_entry@@Base+0x1ab918>
+ 320ae: 48 8d 3d 13 d6 1d 00 lea 0x1dd613(%rip),%rdi # 20f6c8 <SDL_DYN...
```

This confirmed the suspicion that was already noted in Debian, that it likely is around `SDL_DYNAPI_entry`.
Although I'm not a perl master, changing that did not seem too difficult.
I haven't gone through a full reproducible build recheck, but the output seemed to be comparable across various test builds I had.

```
sha256sum 0*/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0*.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  00/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  01/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  02/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  03/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  04/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  05/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  06/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  07/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  08/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
a2d6f07fe79084dd5948759f91b30df624021069161c4128b94b609c5692a4f7  09/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0.3000.9.objdump
```

I know you have switched to a python script in 3.x series, but applying that to 2.30.x would help maintainers, so I wanted to submit it for your consideration.

**Comment by icculus** [comment]
It's fine if we sort the headers, but I'm confused as to how this changes anything; gendynapi only adds things to the end of a file when new APIs are added. It's not part of the general build, and it already outputs things in a specific order when it is manually run.

Am I missing something obvious?

**Comment by cpaelzer** [comment]
Hi @icculus ,
I was really only trying to treat a symptom based on the info (the changing symbols and the suggestion noted in the Debian project) that I found, but not diving deep into how all of it was interconnected. Therefore I sadly can't tell you why/how exactly it would help. As mentioned I didn't go through a full reproducible build re-check, but had 10 builds with it for other reasons all being the same output now.

>  It's not part of the general build, and it already outputs things in a specific order when it is manually run.

You might be right even likely are as you know more about how these things work together. But I've found non explicitly ordered directory reads surprise me more than once, and therefore I think it is just defensive coding to order it explicitly. To me it seems like a "will not break things, but resolve the current working theory" change. If it entirely fixed reproducible builds it is a win, if not it still would have proven that one needs to look elsewhere.

Since the PR was already merged IMHO we can close this issue until there is a Debian/Ubuntu build with the new that goes through reproducible the builds checks.
